_tests.sh

```
@@ -87,131 +163,269 @@ echo "  - <test>/ok.log(all good)"
-    sha1=$(git rev-parse HEAD)
-    git tag -d v1.0
-    git artifact fetch-tags --sha1 "$sha1"
-} > ${test}/run.log 2>&1 || cat ../${test}/run.log
-eval_testcase
+    } > ${root_folder}/${test}/run.log 2>&1 || { pwd && cat ${root_folder}/${test}/run.log; }
+    eval_testcase
+}
 
+if [[ ${arg_testcase:-} == "" ]]; then 
+    # Dynamically list and call test functions
+    test_functions=($(declare -F | awk '{print $3}' | grep -E '^[0-9]+(\.[0-9]+)?$'))
```

Shellcheck (0.10.0)

[warning] Prefer mapfile or read -a to split command output (or quote to avoid splitting).

(SC2207)